README.md
# webpack-build-notifier

A [Webpack](https://webpack.github.io/) plugin that uses the [node-notifier](https://github.com/mikaelbr/node-notifier) package to display OS-level notifications for Webpack build events.

**webpack-build-notifier** can generate notifications for compilation warnings and errors, as well as notify you when the compilation process is triggered and completes successfully. Take a look at the [Config Options](#config-options) to learn more about what **webpack-build-notifier** can do.

[![NPM version](https://img.shields.io/npm/v/webpack-build-notifier.svg)](https://www.npmjs.org/package/webpack-build-notifier)
[![Build Status](https://travis-ci.com/RoccoC/webpack-build-notifier.svg?branch=master)](https://travis-ci.com/RoccoC/webpack-build-notifier)
[![Coverage Status](https://coveralls.io/repos/github/RoccoC/webpack-build-notifier/badge.svg?branch=master)](https://coveralls.io/github/RoccoC/webpack-build-notifier?branch=master)

<img width="334" alt="webpack-build-notifier-error" src="https://user-images.githubusercontent.com/1934237/28636873-799c8c54-71f4-11e7-8d0c-be15ca823f6e.png"><img width="334" alt="webpack-build-notifier-success" src="https://user-images.githubusercontent.com/1934237/28636881-7f394dd2-71f4-11e7-9148-4dba316a41a8.png">

Are you tired of having to constantly switch between your IDE and terminal window to see whether your latest edits resulted in a failed build? Why didn't your latest changes get [hot-loaded](https://github.com/gaearon/react-hot-loader)? Was there a syntax error or failed unit test? With this plugin, you will always be apprised of build problems without having to keep an eye on your terminal window.

To use, install the webpack-build-notifier package `npm install webpack-build-notifier --save-dev` and add the plugin to your [Webpack configuration file](https://webpack.github.io/docs/configuration.html):

```javascript
// webpack.config.js
const WebpackBuildNotifierPlugin = require('webpack-build-notifier');

module.exports = {
  // ... snip ...
  plugins: [
    new WebpackBuildNotifierPlugin({
      title: "My Webpack Project",
      logo: path.resolve("./img/favicon.png"),
      suppressSuccess: true, // don't spam success notifications
    })
  ],
  // ... snip ...
}
```

Config Options
--------------

#### title
The notification title. Defaults to **_Webpack Build_**.

#### logo
The absolute path to the project logo to be displayed as a content image in the notification. Optional.

#### sound
The sound to play for notifications. Set to false to play no sound. Valid sounds are listedin the node-notifier project, [here](https://github.com/mikaelbr/node-notifier). Defaults to **_Submarine_**.

#### successSound
The sound to play for success notifications. Defaults to the value of the *sound* configuration option. Set to false to play no sound for success notifications. Takes precedence over the *sound* configuration option.

#### warningSound
The sound to play for warning notifications. Defaults to the value of the *sound* configuration option. Set to false to play no sound for warning notifications. Takes precedence over the *sound* configuration option.

#### failureSound
The sound to play for failure notifications. Defaults to the value of the *sound* configuration option. Set to false to play no sound for failure notifications. Takes precedence over the *sound* configuration option.

#### compilationSound
The sound to play for compilation notifications. Defaults to the value of the *sound* configuration option. Set to false to play no sound for compilation notifications. Takes precedence over the *sound* configuration option.

#### onCompileStart
A function which is invoked when compilation starts. Optional. The function is passed one parameter:
* {webpack.compilation.Compilation} compilation - The webpack Compilation instance.
Note that `suppressCompileStart` must be `false`.

#### onComplete
A function which is invoked when compilation completes. Optional. The function is passed two parameters:
* {webpack.compilation.Compilation} compilation - The webpack Compilation instance.
* {CompilationStatus} status - one of 'success', 'warning', or 'error'

#### suppressSuccess
Defines when success notifications are shown. Can be one of the following values:
*  false     - Show success notification for each successful compilation (default).
*  true      - Only show success notification for initial successful compilation and after failed compilations.
*  "always"  - Never show the success notifications.
*  "initial" - Same as true, but suppresses the initial success notification.

#### showDuration
True to show the duration of a successful compilation, otherwise false (default).

#### suppressWarning
True to suppress the warning notifications, otherwise false (default).

#### suppressCompileStart
True to suppress the compilation started notifications (default), otherwise false.

#### activateTerminalOnError
True to activate (focus) the terminal window when a compilation error occurs. Note that this only works on Mac OSX (for now). Defaults to **_false_**. Regardless of the value of this config option, the terminal window can always be brought to the front by clicking on the notification.

#### successIcon
The absolute path to the icon to be displayed for success notifications. Defaults to the included **_./src/icons/success.png_**.

![Success](https://github.com/RoccoC/webpack-build-notifier/blob/master/src/icons/success.png?raw=true "Success")

#### warningIcon
The absolute path to the icon to be displayed for warning notifications. Defaults to the included **_./src/icons/warning.png_**.

![Warning](https://github.com/RoccoC/webpack-build-notifier/blob/master/src/icons/warning.png?raw=true "Warning")

#### failureIcon
The absolute path to the icon to be displayed for failure notifications. Defaults to the included **_./src/icons/failure.png_**.

![Failure](https://github.com/RoccoC/webpack-build-notifier/blob/master/src/icons/failure.png?raw=true "Failure")

#### compileIcon
The absolute path to the icon to be displayed for compilation started notifications. Defaults to the included **_./src/icons/compile.png_**.

![Compile](https://github.com/RoccoC/webpack-build-notifier/blob/master/src/icons/compile.png?raw=true "Compile")

#### formatSuccess
A function which returns a formatted notification message on successful compilation.
This function must return a String.
The default formatter will display "Build successful!".
Note that the message will always be limited to 256 characters.

#### messageFormatter
A function which returns a formatted notification message on error or warning.
The function is passed 4 parameters:
* {Object} error/warning - The raw error or warning object.
* {String} filepath - The path to the file containing the error/warning (if available).
* {CompilationStatus} status - Error or warning
* {number} count - How many errors or warnings were raised

This function must return a String.
The default messageFormatter will display the filename which contains the error/warning followed by the
error/warning message.
Note that the message will always be limited to 256 characters.

#### notifyOptions
Any additional node-notifier options as documented in the [node-notifier documentation](https://github.com/mikaelbr/node-notifier).
Note that options provided here will only be applied to the success/warning/error notifications (not the "compilation started" notification). The title, message, sound, contentImage (logo), and icon options will be ignored, as they will be set via the corresponding WebpackBuildNotifier config options (either user-specified or default).

#### onClick
A function called when the notification is clicked. By default it activates the Terminal application.

#### onTimeout
A function called when the notification times out and is closed. Undefined by default.

TypeScript
----------
This project is written in TypeScript, and type declarations are included. You can take advantage of this if your project's webpack configuration is also using TypeScript (e.g. `webpack.config.ts`).

```javascript
// webpack.config.ts
import * as webpack from 'webpack'
import * as WebpackBuildNotifierPlugin from 'webpack-build-notifier';

const config: webpack.Configuration = {
  // ... snip ...
  plugins: [
    new WebpackBuildNotifierPlugin({
      title: "My Webpack Project",
      logo: path.resolve("./img/favicon.png"),
      suppressSuccess: true, // don't spam success notifications
    })
  ],
  // ... snip ...
};

export default config;
```

Notes
-----
After publishing this package I discovered a couple other similar plugins that are worth looking into:
* [webpack-notifier](https://github.com/Turbo87/webpack-notifier)
* [webpack-error-notification](https://github.com/vsolovyov/webpack-error-notification)

Given the purpose and similarities, this project probably could have been a fork of one of these.

Changelog
---------

View the changelog [here](CHANGELOG.md).


CHANGELOG.md
# Changelog

#### 2.3.0

###### _June 19, 2021_

- Added _formatSuccess_ config option.
- Modified _notifyOptions_ config option to accept a function.
- Misc dependency upgrades for security fixes.

#### 2.2.1

###### _April 30, 2021_

- Fixed MessageFormatter type def; consolidated types between `types.ts` and `index.ts`.

#### 2.2.0

###### _April 28, 2021_

- Add compilation status and error/warning count parameters to _messageFormatter_ ([PR](/../../pull/68)).

#### 2.1.1

###### _December 21, 2020_

- Bumped node-notifier to v9.0.0.

#### 2.1.0

###### _June 23, 2020_

- Upgraded dependencies to latest; fixed warning/error handling for child compilers ([#53](/../../issues/53)).

#### 2.0.1

###### _May 23, 2020_

- Fixed bug where setting _sound_ to _false_ had no effect ([#51](/../../issues/51)).

#### 2.0.0

###### _October 15, 2019_

- Converted to TypeScript, increased test coverage.

#### 1.2.2

###### _October 15, 2019_

- Updated node-notifier to fix [#43](/../../issues/43); added test coverage.

#### 1.2.1

###### _October 15, 2019_

- Added _showDuration_ config option for [#48](/../../issues/48).

#### 1.1.1

###### _August 29, 2019_

- Updated node-notifier dependency to latest version to fix [#45](/../../issues/45).

#### 1.1.0

###### _August 10, 2019_

- Added two new config options: onCompileStart, onComplete.

#### 1.0.3

###### _June 4, 2019_

- Updated dependencies to latest versions to fix high-severity js-yaml vulnerability.

#### 1.0.2

###### _May 3, 2019_

- Corrected _sound_ TS types.

#### 1.0.1

###### _April 16, 2019_

- Updated README to include TS usage example.

#### 1.0.0

###### _April 16, 2019_

- Fixed TypeScript definition per issue [#39](/../../issues/39). This is a breaking change for any consumers using TS; bumped major version.

#### 0.1.32

###### _April 16, 2019_

- Fixed TypeScript definitions per issues [#38](/../../issues/38) and [#39](/../../issues/39).

#### 0.1.31

###### _March 20, 2019_

- Added TypeScript definitions per issue [#35](/../../issues/35).

#### 0.1.30

###### _October 7, 2018_

- Fix for [#29](/../../issues/29) where error occurs on Windows 7 when attempting to register the SnoreToast app ID for notifications.

#### 0.1.29

###### _August 21, 2018_

- Fix for [#30](/../../issues/30), compilation notifications no longer ignore _sound_ option.
- Added new config option: _compilationSound_

#### 0.1.28

###### _June 30, 2018_

- Fix for [#28](/../../issues/28) to allow webpack build to complete while notification is visible. Updated to node-notifier@5.2.1.

#### 0.1.27

###### _June 1, 2018_

- Check that error/warning is not null before formatting it.

#### 0.1.26

###### _May 31, 2018_

- Added two new config options: _notifyOptions_ and _onTimeout_ per [#26](/../../issues/26).

#### 0.1.25

###### _April 18, 2018_

- Updated Webpack 4 watchRun hook to use tapAsync to fix [#25](/../../issues/25).

#### 0.1.24

###### _April 15, 2018_

- Updated to use new hooks API for Webpack 4.

#### 0.1.23

###### _February 21, 2018_

- Updated terminal activation to handle VS Code.

#### 0.1.22

###### _January 16, 2018_

- Merged PR to detect terminal application. [#21](/../../issues/21).

#### 0.1.21

###### _December 13, 2017_

- Reworked previous "fix" to use default SnoreToast AppID for Windows toast notifications. [#20](/../../issues/20).

#### 0.1.19

###### _December 13, 2017_

- Added appName parameter to notify config to resolve issue with notifications not being generated in Windows build >=1709 [#20](/../../issues/20).

#### 0.1.18

###### _November 30, 2017_

- Updated `node-notifier` package version to latest; enforced max message length to 256 to fix [#20](/../../issues/20).

#### 0.1.17

###### _November 6, 2017_

- Added notification hook for webpack "watch-run" compilation event to show notifications when the compilation process has started.
  Added _suppressCompileStart_ and _compileIcon_ configuration options to support this. This notification will not be shown by default;
  set _suppressCompileStart_ to _false_ to enable.

#### 0.1.16

###### _July 25, 2017_

- Updated _suppressSuccess_ configuration option to support "always" and "initial" values.

#### 0.1.15

###### _July 17, 2017_

- Updated webpack icons.

#### 0.1.14

###### _June 14, 2017_

- Added _warningSound_ configuration option.

#### 0.1.13

###### _October 19, 2016_

- Added _messageFormatter_ configuration option to allow custom formatting of notification message.

#### 0.1.12

###### _July 25, 2016_

- Bugfix for #6, more null checking.

#### 0.1.11

###### _July 16, 2016_

- Bugfix for #6; added null check for error messages.

#### 0.1.10

###### _July 14, 2016_

- Added reference to _[strip-ansi](https://www.npmjs.com/package/strip-ansi)_ NPM package to remove CLI color formatting from notifications.

#### 0.1.9

###### _July 5, 2016_

- Added new _onClick_ configuration option to allow for specifying of notification click behavior.

#### 0.1.8

###### _February 17, 2016_

- Added new _successSound_ and _failureSound_ configuration options to allow different sounds depending upon the notification type. The _sound_ configuration is still supported, but these two new options will take precedence.

#### 0.1.7

###### _January 18, 2016_

- Fixed _sound_ configuration option to allow "false" value to disable sound.

#### 0.1.6

###### _December 17, 2015_

- Added _suppressWarning_ configuration option.


tests/test.js
import webpack from "webpack";
import WebpackBuildNotifierPlugin from '../index';
const os = require('os');
const webpackConfig = require('./webpack.config');

describe('WebpackBuildNotifierPlugin export initialization test', () => {
    it('WebpackBuildNotifierPlugin should not undefined', () => {
        expect(WebpackBuildNotifierPlugin).not.toBe(undefined);
    });

    it('WebpackBuildNotifierPlugin should not null', () => {
        expect(WebpackBuildNotifierPlugin).not.toBe(null);
    });

    it('WebpackBuildNotifierPlugin should be function', () => {
        expect(typeof WebpackBuildNotifierPlugin).toBe("function");
    });
});

describe('WebpackBuildNotifierPlugin instance test', () => {
    it('Should create a instance with title', () => {
        const instanceWithTitle = new WebpackBuildNotifierPlugin({
            title: "My Project Webpack Build",
        });

        expect(typeof instanceWithTitle).toBe("object");
    });

    it('Check default values', () => {
        const instanceWithTitle = new WebpackBuildNotifierPlugin({
        });
        expect(typeof instanceWithTitle).toBe("object");
        expect(instanceWithTitle.title).toBe("Webpack Build");
        expect(instanceWithTitle.icon).toBe(undefined);
        expect(instanceWithTitle.sound).toBe("Submarine");
        expect(instanceWithTitle.successSound).toBe("Submarine");
        expect(instanceWithTitle.failureSound).toBe("Submarine");
        expect(instanceWithTitle.compilationSound).toBe("Submarine");
        expect(instanceWithTitle.onCompileStart).toEqual(undefined);
        expect(instanceWithTitle.onComplete).toEqual(undefined);
        expect(instanceWithTitle.suppressSuccess).toBe(false);
        expect(instanceWithTitle.suppressWarning).toBe(false);
        expect(instanceWithTitle.suppressCompileStart).toBe(true);
        expect(instanceWithTitle.activateTerminalOnError).toBe(false);
    });

    it('Message formatter test short message', () => {
        const instance = new WebpackBuildNotifierPlugin({
            title: 'Example Configuration'
        });
        const message = { message: "Some short message" };
        expect(instance.messageFormatter(message, '')).toBe('' + os.EOL + "Some short message");
    });

    it('Message formatter test long message', () => {
        const instance = new WebpackBuildNotifierPlugin({
            title: 'Example Configuration'
        });
        const message = { message: "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum." };
        expect(instance.messageFormatter(message, '')).toBe('' + os.EOL + "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has su");
    });

    it('onCompilationWatchRun should not exit with error', () => {
        const instance = new WebpackBuildNotifierPlugin({
            title: 'Example Configuration'
        });
        expect(instance.onCompilationWatchRun("", () => { })).toBe(undefined);
    });
});

describe("Test Webpack build", () => {
    it('Webpack should not throw any errors', (done) => {
        const options = webpackConfig();
        webpack(options, function (err, stats) {
            expect(err).toBeNull();
            expect(stats.hasErrors()).toBeFalsy();
            done();
        });
    });
    it('Webpack should throw error', (done) => {
        const options = webpackConfig('error.js');
        webpack(options, function (err, stats) {
            expect(err).toBeNull();
            expect(stats.hasErrors()).toBeTruthy();
            setTimeout(done, 5500); // build is blocked for 5 sec on notification
        });
    }, 6000);
})

tests/error.js
/**
 * Transpiling this file will fail since there are two default exports.
 */
export default sampleFunction = () => {
    return "Sample value";
}

export default sampleFunction;

tests/warning.js
const sampleFunction = () => {
  return "Sample value";
}

export default sampleFunction;

src/index.ts
/**
 * @class WebpackBuildNotifierPlugin
 * @extends Object
 * A Webpack plugin that generates OS notifications for build steps using node-notifier.
 */
import path from 'path';
import process from 'process';
import os from 'os';
import notifier from 'node-notifier';
import stripAnsi from 'strip-ansi';
import { exec, execFileSync } from 'child_process';
import { Notification } from 'node-notifier/notifiers/notificationcenter';
import { CompilationResult, Config, CompilationStatus } from './types';
import webpack from 'webpack';

const DEFAULT_ICON_PATH = path.resolve(__dirname, 'icons');

export default class WebpackBuildNotifierPlugin {
  private appName: string | undefined = undefined;
  private buildSuccessful: boolean = false;
  private hasRun: boolean = false;

  // config options
  private title: string = 'Webpack Build';
  private logo?: string;
  private sound: string = 'Submarine';
  private successSound: string;
  private warningSound: string;
  private failureSound: string;
  private compilationSound: string;
  private suppressSuccess: boolean | 'always' | 'initial' = false;
  private suppressWarning: boolean = false;
  private suppressCompileStart: boolean = true;
  private activateTerminalOnError: boolean = false;
  private showDuration: boolean = false;
  private successIcon: string = path.join(DEFAULT_ICON_PATH, 'success.png');
  private warningIcon: string = path.join(DEFAULT_ICON_PATH, 'warning.png');
  private failureIcon: string = path.join(DEFAULT_ICON_PATH, 'failure.png');
  private compileIcon: string = path.join(DEFAULT_ICON_PATH, 'compile.png');
  private onCompileStart?: Config['onCompileStart'];
  private onComplete?: Config['onComplete'];
  private onClick: Config['onClick'] = () => this.activateTerminalWindow;
  private onTimeout?: Config['onTimeout'];
  private formatSuccess?: Config['formatSuccess'];
  private messageFormatter?: Config['messageFormatter'];
  private notifyOptions?: Config['notifyOptions'];

  constructor(cfg?: Config) {
    Object.assign(this, cfg);

    if (this.sound) {
      this.successSound = this.successSound ?? this.sound;
      this.warningSound = this.warningSound ?? this.sound;
      this.failureSound = this.failureSound ?? this.sound;
      this.compilationSound = this.compilationSound ?? this.sound;
    }

    this.registerSnoreToast();

    notifier.on('click', this.onClick!);
    /* istanbul ignore else */
    if (this.onTimeout) {
      notifier.on('timeout', this.onTimeout);
    }
  }

  public apply(compiler: webpack.Compiler): void {
    if (compiler.hooks && compiler.hooks.watchRun && compiler.hooks.done) {
      // for webpack >= 4
      /* istanbul ignore else */
      if (!this.suppressCompileStart) {
        compiler.hooks.watchRun.tapAsync('webpack-build-notifier', this.onCompilationWatchRun);
      }
      compiler.hooks.done.tap('webpack-build-notifier', this.onCompilationDone);
    } else {
      // for webpack < 4
      /* istanbul ignore else */
      if (!this.suppressCompileStart) {
        compiler.plugin('watch-run', this.onCompilationWatchRun);
      }
      compiler.plugin('done', this.onCompilationDone);
    }
  }

  private readonly activateTerminalWindow = (): void => {
    if (process.platform === 'darwin') {
      // TODO: is there a way to translate $TERM_PROGRAM into the application name
      // to make this more flexible?
      exec('TERM="$TERM_PROGRAM"; ' +
        '[[ "$TERM" == "Apple_Terminal" ]] && TERM="Terminal"; ' +
        '[[ "$TERM" == "vscode" ]] && TERM="Visual Studio Code"; ' +
        'osascript -e "tell application \\"$TERM\\" to activate"');
    } else if (process.platform === 'win32') {
      // TODO: Windows platform
    }
  };

  // formats the error/warning message
  private readonly formatMessage = (
    error: CompilationResult,
    filepath: string,
    status: CompilationStatus,
    errorCount: number
  ): string => {
    let message: string | undefined = undefined;
    if (this.messageFormatter) {
      message = this.messageFormatter(error, filepath, status, errorCount);
    } else {
      message = (error.message || error.details);
      if (message && error.module && error.module.resource) {
        message = `${filepath}${os.EOL}${message!.replace(error.module.resource, '')}`;
      }
    }

    if (message === undefined) {
      return 'Unknown';
    } else if (typeof message === 'string') {
      return message.substr(0, 256); // limit message length to 256 characters, fixes #20
    } else {
      throw `Invalid message type '${typeof message}'; messageFormatter must return a string.`;
    }
  };

  private readonly onCompilationDone = (results: webpack.Stats): void => {
    let notify: boolean = false;
    let title = `${this.title} - `;
    let msg = this.formatSuccess?.() ?? 'Build successful!';
    let icon = this.successIcon;
    let sound = this.successSound;
    let compilationStatus = CompilationStatus.SUCCESS;

    if (results.hasErrors()) {
      const error = this.getFirstWarningOrError(results.compilation, 'errors');
      const errorFilePath = error.module && error.module.resource ? error.module.resource : '';
      notify = true;
      compilationStatus = CompilationStatus.ERROR;
      title += 'Error';
      msg = this.formatMessage(
        error,
        errorFilePath,
        compilationStatus,
        this.getWarningOrErrorCount(results.compilation, 'errors')
      );
      icon = this.failureIcon;
      sound = this.failureSound;
      this.buildSuccessful = false;
    } else if (!this.suppressWarning && results.hasWarnings()) {
      const warning = this.getFirstWarningOrError(results.compilation, 'warnings');
      const warningFilePath = warning.module && warning.module.resource ? warning.module.resource : '';
      notify = true;
      compilationStatus = CompilationStatus.WARNING;
      title += 'Warning';
      msg = this.formatMessage(
        warning,
        warningFilePath,
        compilationStatus,
        this.getWarningOrErrorCount(results.compilation, 'warnings')
      );
      icon = this.warningIcon;
      sound = this.warningSound;
      this.buildSuccessful = false;
    } else {
      title += 'Success';
      if (this.showDuration) {
        msg += ` [${results.endTime! - results.startTime!} ms]`;
      }
      /* istanbul ignore else */
      if (this.suppressSuccess === 'always' || (this.suppressSuccess === 'initial' && !this.hasRun)) {
        notify = false;
      } else if (this.suppressSuccess === false || !this.buildSuccessful) {
        notify = true; // previous build failed, let's show a notification even if success notifications are suppressed
      }
      this.buildSuccessful = true;
    }

    const notifyOptions =
      (typeof this.notifyOptions === 'function'
        ? this.notifyOptions(compilationStatus)
        : this.notifyOptions) ?? {};

    /* istanbul ignore else */
    if (notify) {
      notifier.notify(
        Object.assign(notifyOptions, {
          title,
          sound,
          icon,
          appName: this.appName,
          message: stripAnsi(msg),
          contentImage: this.logo,
          wait: !this.buildSuccessful
        })
      );
      /* istanbul ignore else */
      if (this.onComplete) {
        this.onComplete(results.compilation, compilationStatus);
      }
    }

    /* istanbul ignore else */
    if (this.activateTerminalOnError && !this.buildSuccessful) {
      this.activateTerminalWindow();
    }

    this.hasRun = true;
  };

  private readonly onCompilationWatchRun = (
    compiler: webpack.compiler.Compiler,
    callback: Function
  ): void => {
    notifier.notify({
      appName: this.appName,
      title: this.title,
      message: 'Compilation started...',
      contentImage: this.logo,
      icon: this.compileIcon,
      sound: this.compilationSound
    } as Notification);
    /* istanbul ignore else */
    if (this.onCompileStart) {
      this.onCompileStart(compiler);
    }
    callback();
  };

  private readonly registerSnoreToast = (): void => {
    // ensure the SnoreToast appId is registered, which is needed for Windows Toast notifications
    // this is necessary in Windows 8 and above, (Windows 10 post build 1709), where all notifications must be generated
    // by a valid application.
    // see: https://github.com/KDE/snoretoast, https://github.com/RoccoC/webpack-build-notifier/issues/20
    /* istanbul ignore else */
    if (process.platform === 'win32') {
      const versionParts = os.release().split('.');
      const winVer = +(`${versionParts[0]}.${versionParts[1]}`);
      /* istanbul ignore else */
      if (winVer >= 6.2) {
        // Windows version >= 8
        const snoreToast = path.join(
          require.resolve('node-notifier'),
          '../vendor/snoreToast',
          `snoretoast-${process.arch === 'x64' ? 'x64' : 'x86'}.exe`
        );
        try {
          execFileSync(
            snoreToast,
            [
              '-appID',
              'Snore.DesktopToasts',
              '-install',
              'SnoreToast.lnk',
              snoreToast,
              'Snore.DesktopToasts'
            ]
          );
          this.appName = 'Snore.DesktopToasts';
        } catch (e) {
          console.error('An error occurred while attempting to install the SnoreToast AppID!', e);
        }
      }
    }
  };

  private readonly getFirstWarningOrError = (
    compilation: webpack.compilation.Compilation,
    type: 'warnings' | 'errors'
  ): any => {
    /* istanbul ignore else */
    if (compilation.children && compilation.children.length) {
      for (let i = 0; i < compilation.children.length; i++) {
        const warningsOrErrors = compilation.children[i][type];
        /* istanbul ignore else */
        if (warningsOrErrors && warningsOrErrors[0]) {
          return warningsOrErrors[0];
        }
      }
    }
    return compilation[type][0];
  }

  private readonly getWarningOrErrorCount = (
    compilation: webpack.compilation.Compilation,
    type: 'warnings' | 'errors',
  ): number => {
    /* istanbul ignore else */
    if (compilation.children && compilation.children.length) {
      const count = compilation.children.reduce(
        (acc, child) => {
          let currentCount = acc;
          const warningsOrErrors = child[type];
          /* istanbul ignore else */
          if (warningsOrErrors) {
            currentCount += warningsOrErrors.length;
          }
          return currentCount;
        },
        0,
      );

      /* istanbul ignore else */
      if (count > 0) {
        return count;
      }
    }
    return compilation[type].length;
  };
}

module.exports = WebpackBuildNotifierPlugin;
